Dispose services in reverse creation order

pakrym opened the pull request:
@davidfowl

pakrym committed:
Dispose services in reverse creation order

davidfowl commented on src/Microsoft.Extensions.DependencyInjection/ServiceLookup/CallSiteExpressionBuilder.cs:
```diff
@@ -164,34 +164,42 @@ protected override Expression VisitScoped(ScopedCallSite callSite, ParameterExpr
-                resolvedExpression);
+                resolvedVariable);
+
+            Expression captureDisposableExpression = Expression.Invoke(
```
Errr?

halter73 replied:
We never tested the actual performance in terms of op/s, just memory allocations. If you look closely at your [original PR](https://github.com/aspnet/DependencyInjection/pull/291) you'll notice that the memory allocations measured are mostly related ServiceProvider.CreateScope()/ctor, and there are no measurements related to the actual capturing disposables of any kind.

If you're really concerned memory you could store some sort of ordering metadata with the captured services (a la Autofac) or pool the Dictionary/List like @benaadams [suggested in the original PR](https://github.com/aspnet/DependencyInjection/pull/291#issuecomment-140784285). Both sound kinda complicated and might even hurt perf in some scenarios.

It might be a good idea to add a BenchmarkDotNet scenario that covers capturing disposables.

davidfowl replied:
Yes, we need measurements. Also `ConcurrentBag` did get way less allocatey in .NET Core 2.0 so maybe it's worth a remeasure.

pakrym replied:
This change would only add scoped and singletons to that list, there aren't many singletons and scope objects are short-lived so I am not super worried about this.

davidfowl replied:
I still want measurements.

davidfowl replied:
Also, all of this work to elide calls to CaptureDisposable this should be repeated here https://github.com/aspnet/DependencyInjection/blob/dev/src/Microsoft.Extensions.DependencyInjection/ServiceLookup/CallSiteExpressionBuilder.cs#L150.

halter73 commented on src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs:
```diff
@@ -655,5 +655,30 @@ public static TheoryData ServiceContainerPicksConstructorWithLongestMatchesData
+            serviceCollection.AddSingleton<IFakeMultipleService, FakeDisposableCallbackInnerService>();
+            serviceCollection.AddSingleton<IFakeMultipleService, FakeDisposableCallbackInnerService>();
+            serviceCollection.AddSingleton<IFakeMultipleService, FakeDisposableCallbackInnerService>();
+            serviceCollection.AddSingleton<IFakeService, FakeDisposableCallbackInnerService>();
```
Nit: might make sense to have a test that mixes service lifetimes. Could be relevant if we go back to tracking disposables using multiple collections.

halter73 commented on src/Microsoft.Extensions.DependencyInjection.Specification.Tests/Fakes/FakeService.cs:
```diff
@@ -2,9 +2,61 @@
 
 namespace Microsoft.Extensions.DependencyInjection.Specification.Fakes
 {
+    public class FakeDisposeCallback
```
There's already 1000 .cs files in Specification.Tests/Fakes, just put all the public classes in their own file.

halter73 commented on src/Microsoft.Extensions.DependencyInjection/ServiceLookup/CallSiteExpressionBuilder.cs:
```diff
@@ -164,34 +164,42 @@ protected override Expression VisitScoped(ScopedCallSite callSite, ParameterExpr
+                    Expression.Block(
+                        assignExpression,
+                        addValueExpression,
+                        captureDisposableExpression)),
```
Suuuper nit: match ordering with runtime resolver. I know it doesn't really matter.

halter73 reviewed (changes requested):
A benchmark would be nice to alleviate @davidfowl's concerns.

pakrym committed:
Add more tests

Eliminate redundant CaptureDisposable call from expression trees
Split fakes into separate files
